### PR TITLE
KFSPTS-14536: Improve error handling for create accounting documents batch job.

### PR DESCRIPTION
**Pull request by nkimble**

### CODE REVIEW

**Comment by nkimble** [comment]
This is ready for code review. While reviewing, please consider what could be causing Quartz to return batch job "success" when the batch job correctly processes bad data and reports a file header failure that should cause Quartz to report the batch job as "failed" due to encountering a file with header problems even though all of the files submitted were processed. This may just be an anomaly that we have encountered in the past but I would like your opinion on whether there may be another underlying issue.

File: src/main/java/edu/cornell/kfs/fp/batch/CreateAccountingDocumentReportItemDetail.java [hunk 1]
```diff
@@ -66,4 +72,30 @@ public void setSuccessfullyRouted(boolean successfullyRouted) {
...
+    public String getRawDocumentData() {
+        return rawDocumentData;
+    }
+
+    public void setRawDocumentData(String rawDocumentData) {
+        this.rawDocumentData = rawDocumentData;
+    }
+    
+    public void appendErrorMessageToExistingErrorMessage(String additionalErrorMessageContent) {
+        this.setErrorMessage((ObjectUtils.isNotNull(this.getErrorMessage()) ? this.getErrorMessage() : KFSConstants.EMPTY_STRING)
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/CreateAccountingDocumentReportItemDetail.java]: To keep the formatting correct (i.e. not have an extra space at the front of the first error message), I think this does need to be an empty string.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentReportServiceImpl.java [hunk 2]
```diff
@@ -4,8 +4,10 @@
 import java.util.List;
 
 import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentReportServiceImpl.java]: Fixed.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java [hunk 3]
```diff
@@ -102,44 +123,59 @@ protected void processAccountingDocumentFromXml(String fileName, CreateAccountin
         } else {
             logReport.getFilesSuccessfullyProcessed().add(fileName);
         }
-        
+        LOG.info("processAccountingDocumentFromXml: Value of reportItem.isNonBusinessRuleFailure just prior to method return =" + reportItem.isNonBusinessRuleFailure() + "=");
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java]: Removed.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentValidationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import edu.cornell.kfs.fp.batch.service.CreateAccountingDocumentValidationService;
+import edu.cornell.kfs.fp.batch.xml.AccountingXmlDocumentEntry;
+import edu.cornell.kfs.fp.batch.xml.AccountingXmlDocumentListWrapper;
+
+public class CreateAccountingDocumentValidationServiceImpl implements CreateAccountingDocumentValidationService {
+    private static final Logger LOG = LogManager.getLogger(CreateAccountingDocumentValidationServiceImpl.class);
+    
+    protected ConfigurationService configurationService;
+    
+    public boolean isValidXmlFileHeaderData(AccountingXmlDocumentListWrapper accountingXmlDocument, CreateAccountingDocumentReportItem reportItem) {
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentValidationServiceImpl.java]: Added.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentValidationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        boolean allValidationChecksPassed = isReportEmailAddressValid(accountingXmlDocument.getReportEmail(), reportItem);
+        allValidationChecksPassed &= isOverviewValid(accountingXmlDocument.getOverview(), reportItem);
+        return allValidationChecksPassed;
+    }
+    
+    public boolean isAllRequiredDataValid(AccountingXmlDocumentEntry accountingXmlDocument, CreateAccountingDocumentReportItemDetail reportItemDetail) {
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentValidationServiceImpl.java]: Added.

File: src/main/java/edu/cornell/kfs/fp/batch/xml/AccountingXmlDocumentEntry.java [hunk 6]
```diff
@@ -12,6 +12,7 @@
 import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
 
 import org.apache.commons.lang.StringUtils;
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/AccountingXmlDocumentEntry.java]: Updated.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java [hunk 7]
```diff
@@ -70,27 +71,47 @@ public boolean createAccountingDocumentsFromXml() {
 
     protected void processAccountingDocumentFromXml(String fileName, CreateAccountingDocumentLogReport logReport) {
         CreateAccountingDocumentReportItem reportItem = new CreateAccountingDocumentReportItem(fileName);
+        boolean headerValidationFailed = true;
```
[the remark this answers is not in the document]
Comment (nkimble) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImpl.java]: Reversed the logic for failure as requested but please keep in mind, should the validation hiccup for some reason, the job will attempt to continue to process the file and create accounting documents  with bad header data email addresses or missing overview data...just as it does today.

File: src/main/java/edu/cornell/kfs/fp/batch/CreateAccountingDocumentReportItemDetail.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/CreateAccountingDocumentReportItemDetail.java]: OK, thanks for checking

**Comment by nkimble** [comment]
I have added unit tests for the validation. There is now full coverage with 16 unit tests representing all the combinations for having each of the four items with bad data.  This is ready to be reviewed again.

**Pull request merged**